flux-middle/src/rty/subst.rs

```diff
@@ -329,10 +330,44 @@ impl TypeFolder for GenericsSubstFolder<'_> {
         }
     }
 
+    /*  [NOTE:index-subst]
+
+    Consider
+
+    ```rust
+    fn choose <T as base>(b: bool, x: T[@n], y: T[@m]) -> T[if b { n } else { m }])
+    ```
+
+    and then a client
+
+    ```rust
+    pub fn test01() {
+        assert(choose(true, 0, 1) == 0);
+    }
+    ```
+
+    At the callsite `choose(true, 0, 1)` there are *two* substitutions going on
+    in the signature for `choose` i.e. for the `T[@n]` and `T[@m]`
+
+    1. `T` -> `i32{v: ...}`
+    2. earlybound `n, m` -> fresh-evars.
+
+    The trouble is that if you solely rely on the `bty_for_param` code below,
+    then the `n, m` substitutions "get lost" and so those variables are not
+    "solved for" and you get that pesky instantiation error.
+
+    Instead, we _first_ substitute for the indices, and then let `bty_for_param`
+    do its business.
+
+    */
```

Comment:
Prefer line comments over block comments per the style guidelines https://rust-coding-guidelines.github.io/rust-coding-guidelines/safe-guides/code_style/comments/P.CMT.03.html